Add java 21 variant to GitHub actions build

dlmarion opened the pull request.

dlmarion committed:
Add java 21 variant to GitHub actions build

dlmarion committed:
Add java 21 variant to GitHub actions build

dlmarion committed:
Merge branch 'add-java-21-build' of github.com:dlmarion/accumulo into add-java-21-build

dlmarion commented:
Failed for 2.1 build, switched to 3.1

dlmarion commented:
Failed on 2.1 branch with build errors in the VFS ClassLoader. On the 3.1 branch it's failing on Key.java.

ctubbsii commented on .github/workflows/maven.yaml:
```diff
@@ -68,19 +68,20 @@ jobs:
-          - {name: 'compat',        javaver: 11, args: 'package -DskipTests -Dversion.hadoop=3.0.3 -Dversion.zookeeper=3.5.10'}
-          - {name: 'errorprone',    javaver: 11, args: 'verify -Perrorprone,skipQA'}
-          - {name: 'jdk17',         javaver: 17, args: 'verify -DskipITs'}
+          - {name: 'unit-tests',    distribution: adopt,  javaver: 11, args: 'verify -PskipQA -DskipTests=false'}
```
To be consistent with the `javaver` param, maybe `distribution` should be called `javadist`?

ctubbsii reviewed (changes requested):
I'm in favor of the overall change (minor suggestion on the name of the matrix parameter, but that's not important).

However, I'd prefer to not see constant failures on our builds for every PR. So, it would probably make sense to fix the outstanding issues before applying this change. While the build stopped at Key.java, there's a good chance there are other issues once it gets past that one.

dlmarion committed:
Modifications for JDK21 compatibility

Minor code modifications for the following errors:

possible 'this' escape before subclass is fully initialized
non-transient instance field of a serializable class declared with a non-serializable type
implicit cast from long to int in compound assignment is possibly lossy

dlmarion commented:
I fixed some of the errors in 99e7ecd. Currently at the point where tests using PowerMock are failing. I think PowerMock might need to be ripped out at this point.

keith-turner commented:
> Currently at the point where tests using PowerMock are failing. I think PowerMock might need to be ripped out at this point.

What problem is powermock causing w/ jdk21?  Like does it cause unit test to fail or a maven plugin to fail.

dlmarion commented:
> > Currently at the point where tests using PowerMock are failing. I think PowerMock might need to be ripped out at this point.
> 
> What problem is powermock causing w/ jdk21? Like does it cause unit test to fail or a maven plugin to fail.

```
[ERROR] org.apache.accumulo.coordinator.CompactionCoordinatorTest.testCoordinatorColdStartNoCompactions -- Time elapsed: 1.460 s <<< ERROR!
java.lang.IllegalStateException: Failed to transform class with name org.apache.accumulo.coordinator.CompactionCoordinatorTest$TestCoordinator. Reason: Cannot read field "string" because "utf" is null
	at org.powermock.core.classloader.javassist.JavassistMockClassLoader.defineAndTransformClass(JavassistMockClassLoader.java:119)
	at org.powermock.core.classloader.MockClassLoader.loadMockClass(MockClassLoader.java:174)
	at org.powermock.core.classloader.MockClassLoader.loadClassByThisClassLoader(MockClassLoader.java:102)
	at org.powermock.core.classloader.DeferSupportingClassLoader.loadClass1(DeferSupportingClassLoader.java:147)
	at org.powermock.core.classloader.DeferSupportingClassLoader.loadClass(DeferSupportingClassLoader.java:98)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:526)
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:421)
	at java.base/java.lang.Class.forName(Class.java:412)
	at javassist.runtime.Desc.getClassObject(Desc.java:72)
	at javassist.runtime.Desc.getClassType(Desc.java:181)
	at javassist.runtime.Desc.getType(Desc.java:151)
	at javassist.runtime.Desc.getType(Desc.java:107)
	at org.apache.accumulo.coordinator.CompactionCoordinatorTest.testCoordinatorColdStartNoCompactions(CompactionCoordinatorTest.java:231)
```

Looking at the PowerMock mailing list and GitHub, I think it might be dead. Last commit back in Feb 2022, last release in 2020.

keith-turner commented:
> Looking at the PowerMock mailing list and GitHub, I think it might be dead. Last commit back in Feb 2022, last release in 2020.

That is a good find from trying to build with 21

EdColeman commented:
[Mockito](https://site.mockito.org/) may also be an alternative.

dlmarion commented:
> [Mockito](https://site.mockito.org/) may also be an alternative.

I looked at their site and did not get a good feel that it would work with Java 21 at this point either.

dlmarion commented:
Created #3817 to discuss removing the tests and PowerMock.

ctubbsii commented on core/src/main/java/org/apache/accumulo/core/security/Authorizations.java:
```diff
@@ -43,11 +43,11 @@
   private static final long serialVersionUID = 1L;
-  private static final Set<ByteSequence> EMPTY_AUTH_SET = Collections.emptySet();
-  private static final List<byte[]> EMPTY_AUTH_LIST = Collections.emptyList();
+  private static final HashSet<ByteSequence> EMPTY_AUTH_SET = new HashSet<>();
+  private static final ArrayList<byte[]> EMPTY_AUTH_LIST = new ArrayList<>();
```
These changes will make these mutable, and they could leak out, have stuff added to them, and then no longer be empty.

I'm not sure what's motivating this particular change, but it seems like it might have been okay before? I'm not sure why this one would be a problem for JDK 21.

keith-turner replied:
It would be good to keep those immutable.

dlmarion replied:
The error was `non-transient instance field of a serializable class declared with a non-serializable type`

keith-turner replied:
Could we delete those two variable and wherever they are used use Set.of() and List.of() instead?

ctubbsii replied:
`Set.of()` and `List.of()` have issues with `.contains(null)`, so can be annoying (they throw an exception rather than return false). They might work here if they are serializable, but I'm not sure.

I'm very confused about the error, because these should be static fields, not instance fields, so shouldn't be serialized with the instance anyway. But, if I'm confused about that, it seems the simplest fix is to just mark them transient, so they are excluded from serialization. They are already initialized when the class is loaded, so it shouldn't be an issue.

dlmarion replied:
So, it wasn't the static fields that were an issue. It was the instance fields. But then we return the static fields in some cases. I may have over-corrected on this one.

keith-turner replied:
> However, I'm very confused about the error, because these should be static fields, not instance fields, so shouldn't be serialized with the instance anyway. 

I was wondering about that, if you find out the reasoning let me know.

> We could certainly get rid of the variables and just use the emptyList or emptySet instead, though.

That sounds good, no surprises with null.

ctubbsii commented on pom.xml:
```diff
@@ -783,6 +783,7 @@
             <showWarnings>true</showWarnings>
             <compilerArgs>
               <arg>-Xlint:all</arg>
+              <arg>-Xlint:-this-escape</arg>
```
I don't think this is a good one to ignore. This one catches problems of improperly constructed classes calling methods from uninitialized subclasses while the parent class is in the process of being constructed. These shouldn't be ignored, they should be fixed.

keith-turner replied:
Could add suppression in the code of the places it happens now and open follow on issues, but not suppress it everywhere.

ctubbsii replied:
I'd rather just fix them first. It seems like those could be legitimate bugs. There's no rush to get this in. I'd like to understand the scope of the problem first, though. If you'll indulge me and hold off, I could spend some time trying to track these down and knock them out next week.

dlmarion commented:
I'm wondering if we should just close this PR and punt on Java 21 source compatibility for now.

ctubbsii commented:
> I'm wondering if we should just close this PR and punt on Java 21 source compatibility for now.

Can just leave it open until we work through the issues. No need to close it. It's useful to track the status.

keith-turner commented:
> I'm wondering if we should just close this PR and punt on Java 21 source compatibility for now.

It feels like this JDK version has improved the static analysis it does, so maybe there is some benefit to running it as part of the build.

dlmarion commented:
This pull request is OBE. Closing.
